### ISSUE TITLE
Enhance `reportInvalidTypeVarUse` to report situations where function-scoped TypeVar may go unsolved

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#3495]
The `reportInvalidTypeVarUse` diagnostic check reports cases where a TypeVar appears only once in a generic function signature. 

It may be worthwhile also flagging the case where a TypeVar appears multiple times but appears only within input parameters that have associated default argument values and those values are specified as `...`. This can occur in type stubs, for example. It can lead to TypeVars that remain unsolved or are solved as `Unknown`.

Here's an example of this: https://github.com/microsoft/pyright/issues/3494

**Comment by erictraut** [comment]
I've implemented this check but found that it was quite noisy. In many cases, it is assumed that a potentially-unsolved TypeVar can be included in the return type if it is part of a union with other types. Mypy and pyright (and probably other type checkers) eliminate the unsolved TypeVar from the union when evaluating the return type. For example:

```python
_T = TypeVar("_T")
def foo(x: _T = ...) -> _T | int: ...

reveal_type(foo("")) # str | int
reveal_type(foo()) # int
```

So I've made the check smart enough not to report an error in this case. Without this exemption, it is very noisy.

With this exemption, it's still a bit noisy for typeshed. It found 10 cases in the stdlib stubs and 30 additional cases in third-party stubs. In most of these cases, the remedy is trivial and simply involves deleting an unnecessary `= ...` from an existing overload signature. In other cases, an overload will need to be added.

@JelleZijlstra, based on this level of noise, do you still feel good about the feature?

**Comment by JelleZijlstra** [comment]
Thanks! I do think it's helpful:

- We had two issues in pyright recently where people encountered examples of this problem
- The error I believe shows up only with strict typechecking options enabled, so probably there are more people who are just silently getting Any.

Would you mind opening an issue in typeshed with the list of issues? I'll try to fix at least a few, and hopefully others will help too.

**Comment by JelleZijlstra** [comment]
Minor feedback: The error message is `Type variable "_ArgumentParserT" may go unsolved if argument is not provided for parameter with default value`. Would it make sense to include the name of the parameter? The line and column number are technically enough, but including the parameter name would be a bit user-friendlier.

**Comment by JelleZijlstra** [comment]
I noticed that the `__init__` method of `fileinput.FileInput` is also affected by this issue, but isn't caught by your check: https://github.com/python/typeshed/blob/8a07787f48b4cf3c46fcb9107c5b31388b7b32d2/stdlib/fileinput.pyi#L68. At the risk of getting myself more work, is that easy to catch too?

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.249, which I just published. It will also be included in the next release of pylance.
